[browser][mt] Update memory views after growth, refactor string processing, fix SharedArrayBuffer detection

IsaMorphic opened the pull request:
"memory pooling or reclamation mechanism that they use internally that causes SharedArrayBuffer to cross the iframe boundary"

- test for `SharedArrayBuffer` via `buffer[Symbol.toStringTag] === "SharedArrayBuffer"`
    -  @IsaMorphic adapted a fix to this issue from this PR in the emscripten repository: https://github.com/emscripten-core/emscripten/pull/16994
- added `copyBufferIfNecessary` helper
   - the method is NOOP in non-threaded build
- added `receiveWorkerHeapViews` 
   - the method is NOOP in non-threaded build
- replaced `Module.HEAP`xx with new `localHeapViewXXX()`
   - which calls `receiveWorkerHeapViews` and produces updated short lived "local" view
- added `-s TEXTDECODER=0` emcc linker flag for MT build
- wrap `Module.UTF8ArrayToString` and `Module.UTF8ToString` with our own `utf8ToString` and `decodeUTF8`
   - this will copy the bytes and use `TextDecoder` for longer strings instead of emscripten byte-by-byte loop
- move code into `encodeUTF8` and `utf8ToStringRelaxed` of `strings.ts`
- prefer `HEAPU8` over `HEAP8` for code consistency
- `string.ts` refactoring

Fixes https://github.com/dotnet/runtime/issues/86696
Fixes https://github.com/dotnet/runtime/issues/86642

Related issues: 
https://github.com/emscripten-core/emscripten/issues/15217

IsaMorphic committed:
Fix crashing bug in StringDecoder.decode when accessing SharedArrayBuffer from different JavaScript realm

When performing tasks such as allocating or marshalling large amounts of memory in multithreaded .NET land, inconsistent and frequent crashing behavior is exhibited. 

Related issues: 
https://github.com/emscripten-core/emscripten/issues/15217
https://github.com/dotnet/aspnetcore/issues/48390

IsaMorphic commented:
@dotnet-policy-service agree

kg commented on src/mono/wasm/runtime/strings.ts:
```diff
@@ -77,7 +77,10 @@ export class StringDecoder {
+        
+            // BEWARE: In some cases, `instanceof SharedArrayBuffer` returns false even though buffer is an SAB.
+            // Patch adapted from https://github.com/emscripten-core/emscripten/pull/16994
+            const subArray = typeof SharedArrayBuffer !== "undefined" && Object.prototype.toString.call(Module.HEAPU8.buffer) === "[object SharedArrayBuffer]"
```
This toString operation and comparison will probably add measurable overhead to the cost of decoding strings, so I'd like to see a better solution

IsaMorphic replied:
Good catch! Using the object's `[Symbol.toStringTag]` property should do the trick, since that's what Object.toString uses under the hood.

kg commented:
Thanks for your contribution! If I understand correctly, this instanceof bug affects scenarios involving iframes because the iframe is a different realm and has its own builtins like Array and SharedArrayBuffer, right? So any case where we are using instanceof is potentially broken in that case, which is what motivates APIs like Array.isArray.

IsaMorphic committed:
Reimplementing StringDecoder.decode fix using same principle but this time avoiding call to Object.toString

kg commented on src/mono/wasm/runtime/strings.ts:
```diff
@@ -77,7 +77,11 @@ export class StringDecoder {
+            // BEWARE: In some cases, `instanceof SharedArrayBuffer` returns false even though buffer is an SAB.
+            // Patch adapted from https://github.com/emscripten-core/emscripten/pull/16994
+            // See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag
+            const subArray = typeof SharedArrayBuffer !== "undefined" && Module.HEAPU8.buffer[Symbol.toStringTag] === "SharedArrayBuffer"
```
toStringTag is the perfect solution for this I think 👍

IsaMorphic replied:
I think so too! Very clean and just as easy to write.

pavelsavara commented:
How could dotnet wasm linear memory come from iframe ? 
Are we talking about 2 runtimes sharing memory across the iframe boundary ?

I would appreciate small demo project which reproduces the issue before we merge this. 
Adapting one of the [samples](https://github.com/dotnet/runtime/tree/main/src/mono/sample/wasm) would be ideal.

Possibly related issue https://github.com/dotnet/runtime/issues/86642

IsaMorphic commented:
@pavelsavara 
To answer your question, yes: in my implementation specific use case, I am using a Blazor app as an embeddable image renderer to demonstrate a future product. In my application I am using 2 or more runtimes per page (each in their own iframe). However, it seems (from the related emscripten bug) that all modern browsers might have some type of memory pooling or reclamation mechanism that they use internally that causes buffers to cross the iframe boundary. Though it still could be misuse on the part of the .net runtime, as well :P

Please see the latest commit on my fork's `broken-sample` branch. Use the typical launch commands for the `browser-threading-minimal` sample, but make sure to browse to `http://localhost:<port>/embeds.html` and read the text on the page. This updated sample replicates the issue extremely consistently. 

I have not tried to see if my patch helps, as I have not build the runtime with the `-sTEXTDECODER=0` flag and tested it in this way yet. 

Hope this is helpful.

pavelsavara commented:
`UTF8ArrayToString` is emscripten method and the issue is the same https://github.com/emscripten-core/emscripten/issues/18034
`TEXTDECODER=0` should have fixed that for you.

"memory pooling or reclamation mechanism that they use internally that causes buffers to cross the iframe boundary" could you please point me to more details about it ?

pavelsavara commented:
Also, if there is buffer reuse across iframes, it should also manifest without threads, right ?
We have iframe based [benchmark](https://github.com/dotnet/runtime/blob/main/src/mono/sample/wasm/browser-bench/AppStart.cs) which we run daily. I don't remember such issue to manifest there. 

@radekdoulik would we capture/notice the error if one of the AppStart measures failed to start the runtime in the iframe ?

pavelsavara commented:
I get it now, in our benchmark it's not a `SharedArrayBuffer`. I guess we should create MT flavor of the benchmark soon anyway.

pavelsavara commented:
> Please see the latest commit on my fork's `broken-sample` branch. 

https://github.com/dotnet/runtime/compare/main...IsaMorphic:runtime-wasmfix:broken-sample

I will have look soon, probably next week. Thanks a lot!

I would still love to read about how this is not a browser bug. :-D

pavelsavara committed:
Merge branch 'main' into IsaMorphic_main

pavelsavara commented:
I used your sample and indeed I can reproduce it.
![image](https://github.com/dotnet/runtime/assets/271576/c8025adf-d24b-4443-b838-4213fd2bf33a)

I will push more comprehensive fix into this PR

pavelsavara committed:
- added updateGrowableHeapViews before each usage of Module.HEAPxx

- wrap Module.UTF8ArrayToString and Module.UTF8ToString with our own utf8ToString and decodeUTF8
- move code into encodeUTF8 and utf8ToStringRelaxed
- prefer HEAPU8 over HEAP8 for consistency

pavelsavara commented:
I made bit of the monster from your PR @IsaMorphic 😁 I hope you don't mind

@kg please re-review. I updated the PR description.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix V8

pavelsavara commented:
/azp run runtime-wasm

lambdageek commented on src/mono/wasm/runtime/strings.ts:
```diff
@@ -272,3 +269,68 @@ export function js_string_to_mono_string_new(string: string): MonoString {
+export function copy_string_buffer_as_necessary(view: Uint8Array, start: CharPtr, end: CharPtr): Uint8Array {
+    updateGrowableHeapViews();
+    // this condition should be eliminated by rollup on non-threading builds
+    const needsCopy = mayNeedCopy && view.buffer[Symbol.toStringTag] === "SharedArrayBuffer";
```
Let's make a general purpose utility

```ts
function isSharedArrayBuffer(x: any): x is SharedArrayBuffer
{
  return mayNeedCopy && view.buffer[Symbol.toStringTag] === "SharedArrayBuffer"
}

pavelsavara replied:
This is what `StringDecoder.decode` boiled down to in non-MT build

![image](https://github.com/dotnet/runtime/assets/271576/cc2fd1b4-d829-4ca6-ac44-e25e7e01be9d)

IsaMorphic commented:
This is really awesome guys!! I'm so happy that my simple one-line fix has prompted the implementation of a more robust solution. Great job especially to @pavelsavara 😄

I'm wondering how I could go about using these patches in a Blazor project build? Is there an MSBuild hack that I could use, or do I have to be more patient and wait for the next preview? 

I ask because I want to test this patch on my original project before entirely deciding that the issue is completely fixed.

Please let me know!

maraf commented:
> I'm wondering how I could go about using these patches in a Blazor project build? Is there an MSBuild hack that I could use, or do I have to be more patient and wait for the next preview?

You can try this https://github.com/dotnet/runtime/blob/main/src/mono/wasm/build/README.md#blazor

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -83,25 +89,30 @@ export function setU32_unchecked(offset: MemOffset, value: NumberOrPointer): voi
 }
 
 export function setI32_unchecked(offset: MemOffset, value: number): void {
+    updateGrowableHeapViews();
```
Have you been able to benchmark the impact of this change? I'm concerned it will be a measurable performance regression in single-threaded builds for no upside

pavelsavara replied:
`updateGrowableHeapViews` is NOOP in single threaded, because rollup. I will double check that .js file to make sure.

pavelsavara replied:
This is MT version
![image](https://github.com/dotnet/runtime/assets/271576/ad12876d-f8b4-461e-9c04-6e47dc19e544)

This is single threaded version
![image](https://github.com/dotnet/runtime/assets/271576/a8468300-9eb9-4985-96e8-e32590230053)

The string `"Module.wasmMemory not set"` is not even in the file.

I will remove the `mono_assert` for MT performance sake.

pavelsavara committed:
feedback & cleanup

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
@IsaMorphic to build runtime use 
`build.cmd -os browser -subset mono+libs -c Debug /p:MonoWasmBuildVariant=multithread` and then maraf's advice

pavelsavara commented:
> `Reason: Error: Assert failed: Can't find asset for js-module-runtime`

I think this is because we made changes in the tooling for  preview5 which are not compatible with preview4. Right @maraf ?

pavelsavara committed:
feedback from @kg, optimize away multiple calls to updateGrowableHeapViews() where possible

add localHeapView* as public API

maraf commented:
> > `Reason: Error: Assert failed: Can't find asset for js-module-runtime`
> 
> I think this is because we made changes in the tooling for preview5 which are not compatible with preview4. Right @maraf ?

Yes, this is incompatibility between SDK and runtime pack. Installing newer (nightly) preview will fix it

pavelsavara committed:
few more fixes

pavelsavara committed:
fix

pavelsavara committed:
make strings.ts more readable

pavelsavara committed:
drop legacy string stuff

pavelsavara committed:
Merge branch 'main' into IsaMorphic_main

pavelsavara commented:
I processed further feedback from @kg about not calling `updateGrowableHeapViews()` from tight loops.

While doing that I noticed that we have too many loops where we convert strings with different encodings, so I moved all of it into `string.ts`. 
And while I worked on that I dropped `StringDecoder` class because it was not helping anything. 
I dropped few legacy methods and renamed the rest to camelCase.

@ilonatommy I touched some of the hybrid locale stuff, please have look.

There is bit of risk that this change will introduce perf regression, but it was that already yesterday.
@radekdoulik could you please run benchmark on this before I merge it ?

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

kg commented on src/mono/wasm/runtime/debug.ts:
```diff
@@ -59,6 +61,7 @@ export function mono_wasm_add_dbg_command_received(res_ok: boolean, id: number,
 }
 
 function mono_wasm_malloc_and_set_debug_buffer(command_parameters: string) {
+    const heapU8 = localHeapViewU8();
```
this is incorrect because we're calling malloc after, we need to do this after the malloc call

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -65,43 +68,58 @@ export function setB32(offset: MemOffset, value: number | boolean): void {
+}
+
+// does not check for growable heap
+export function setU16_local(offset: MemOffset, value: number): void {
```
it may make sense for the _local functions to actually accept the heap as an argument, so you can't use them without having previously grabbed it

kg commented on src/mono/wasm/runtime/net6-legacy/js-to-cs.ts:
```diff
@@ -150,10 +150,13 @@ function _extract_mono_obj_root(should_add_in_flight: boolean, js_obj: any, resu
 // https://github.com/Planeshifter/emscripten-examples/blob/master/01_PassingArrays/sum_post.js
 function js_typedarray_to_heap(typedArray: TypedArray) {
+    assert_legacy_interop();
+    const heapU8 = localHeapViewU8();
```
move below malloc

kg commented on src/mono/wasm/runtime/roots.ts:
```diff
@@ -217,7 +217,7 @@ export class WasmRootBufferImpl implements WasmRootBuffer {
         this._check_in_range(index);
         const offset = this.get_address_32(index);
-        return <any>Module.HEAPU32[offset];
+        return <any>localHeapViewU32()[offset];
```
roots and root buffers don't need growth checks; we pre-allocated them so as long as we do a growth check after the initial allocation the addresses will always be visible.

pavelsavara replied:
I put back `Module.HEAPU32` to all root code, correct also for MT ?

pavelsavara replied:
I'm not sure. I think `grow` would disconnect the views and they will be returning `undefined`

kg replied:
Good point

kg commented on src/mono/wasm/runtime/web-socket.ts:
```diff
@@ -336,16 +332,10 @@ function _mono_wasm_web_socket_send_buffering(ws: WebSocketExtension, buffer_vie
-                ? (<any>buffer).slice(0, offset)
-                : buffer.subarray(0, offset);
-            return _text_decoder_utf8.decode(bytes);
+            const bytes = copyBufferIfNecessary(buffer, 0 as any, offset as any);
```
there is type confusion here and in the original code - subarray is a method of ArrayView, which implies buffer is a Uint8Array or other type, not an ArrayBuffer. Which means `buffer instanceof SharedArrayBuffer` would never be true.

I think the operation we actually want here is named something like `ensureViewHasNonSharedBuffer`.

pavelsavara replied:
the `buffer` in WS code is always instance of `Uint8Array` and the code is correct, I will rename `copyBufferIfNecessary` to `viewOrCopy`

pavelsavara replied:
name `buffer` in WS is bit misleading, but most of the time it's dedicated buffer, not view of linear memory.

ilonatommy commented on src/mono/wasm/runtime/hybrid-globalization/change-case.ts:
```diff
@@ -31,20 +27,19 @@ export function mono_wasm_change_case_invariant(src: number, srcLength: number,
         if (!cultureName)
             throw new Error("Cannot change case, the culture name is null.");
-        const input = get_utf16_string(src, srcLength);
+        const input = utf16ToString(src, src + 2 * srcLength);
```
This place used to use loop-only. When we change it to using the TextDecoder on browser, we might have some hybrid tests failing - encoders/decoders behave like NLS, not like ICU.

ilonatommy reviewed (approved):
Globalization pointer math checked.

pavelsavara committed:
feedback

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
Merge branch 'main' into IsaMorphic_main

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
I guess this is that we eliminated bunch of checks in the non-MT string conversion.

![image](https://github.com/dotnet/runtime/assets/271576/47b3616b-b7ae-45ea-b464-e68e37e1cdf2)

https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2023-05-24T10%3A51%3A33.000Z&endDate=2023-05-29T21%3A06%3A42.000Z&tasks=%2CJSInterop
